_posts/2019-11-08-method-references-the-unpolished-gem-of-java-8.markdown

```diff
@@ -0,0 +1,299 @@
+
+### Calling functional interface's methods from the method reference
+
+Sometimes the object's method alone are not enough to implement the logic.
+In these cases, we can take one of the few approaches:
+
+ 1. create a decorator around the object and add missing methods
+ 2. write a new method that takes the object as an argument
+ 3. put the code in-line
+ 4. extend the offending class
+
+I listed them in order of my own preference.
+Each of these has their own advantages and disadvantages.
+
+In many cases, the Functional Interface of our Lambda already provides default methods with necessary logic.
+For example, consider a rather common operation: negating a `Predicate`.
+When I tried it for the first time, my thoughts went something like this:
+
+{% highlight java %}
+dogs.stream()
+    .filter(dog -> !veterinarian.isHealthy(dog))
+    .forEach(sickDog -> veterinarian.treat(sickDog));
+{% endhighlight %}
+
+Alright, I can definitely use a Method Reference in the `forEach`:
+
+{% highlight java %}
+dogs.stream()
+    .filter(dog -> !veterinarian.isHealthy(dog))
+    .forEach(veterinarian::treat);
+{% endhighlight %}
+
+Looks good, now the filter...
+Oh, it doesn't recognize the type?
+Let's try anyway.
+
+{% highlight java %}
+dogs.stream()
+    .filter(((Predicate<Dog>) veterinarian::isHealthy).negate())
+    .forEach(veterinarian::treat);
+{% endhighlight %}
+
+Yup, it looks even worse than before.
+
+To get any further, depending on the version of Java you are using, you can either:
+
+ * use the static `not` method of the `Predicate` class (Java 11), or
+ * write one yourself
+
+If you choose the former, you'll end up with something similar to:
+
+{% highlight java %}
+public static <T> Predicate<T> not(final Predicate<? super T> predicateToNegate) {
+    Objects.requireNonNull(predicateToNegate);
+    return predicateToNegate.negate()::test;
+}
+{% endhighlight %}
+
+This will sit in some `PredicateUtils` class that nobody will remember next time the same problem occurs.
+Anyway, our code is now a bit shorter, but also slightly more difficult to read:
+
+{% highlight java %}
+dogs.stream()
+    .filter(not(veterinarian::isHealthy))
+    .forEach(veterinarian::treat);
+{% endhighlight %}
+
+Except…
+
+> Take all dogs for which not veterinarian says they are healthy
+
+That does not read well at all.
+
+The `java.util.function.Predicate` interface already has a `negate` method.
+It seems logical that negating a Method Reference of a `Predicate` should be possible.
+Then why can we not write something like this instead?
+
+{% highlight java %}
+dogs.stream()
+    .filter(veterinarian::isHealthy.negate())
+    .forEach(veterinarian::treat);
+{% endhighlight %}
+
+Or perhaps by applying static methods like:
+
+{% highlight java %}
+dogs.stream()
+    .filter(veterinarian:not:isHealthy)
+    .forEach(veterinarian::treat);
+{% endhighlight %}
+
+The two examples above are obviously not flawless.
```

Comment:
I actually find them worse than the lambda hehe

Reply:
It's a matter of opinion then, I guess (;
Still better to have more than one option available!